Publicize: add filter to allow user role restriction to be overridden

jb510 opened the pull request:
No existing issue

It would make more sense to to me in general to allow Publicize to appear on the 'edit_posts' capability rather than 'publish_posts'.  We have a site where 'publish_posts' has been removed from the Author role so that Editors can schedule posts.  

#### Changes proposed in this Pull Request:

Rather than retroactively change this however this filter allows for overriding the default when needed.

```
/**
 * Modifies JetPack Publicize Module to required capability for custom user role
 * @return string user role
 */
function jb_publicize_capability() {
	return 'edit_posts';
}
add_filter( 'publicize_capability', 'jb_publicize_capability' );
```

jeherve commented on modules/publicize/ui.php:
```diff
@@ -22,8 +22,9 @@ function __construct() {
-		// Show only to users with the capability required to manage their Publicize connections.
-		if ( ! current_user_can( 'publish_posts' ) ) {
+		// Show only to users with the capability required to manage their Publicize connections.  Filter to allow for custom roles. 
+		$capability = apply_filters( 'publicize_capability', 'publish_posts' );
```
Could you document the filter, like so:
https://github.com/jb510/jetpack/blob/jb510-patch-1/modules/publicize/ui.php#L563

Thanks!

samhotchkiss replied:
Also, let's make the filter name `jetpack_publicize_capability` for better safety 👍

jb510 commented:
Updated.  @jeherve I added inline phpdoc as best I could, but they definitely need to be review by someone experienced with your ways of writing them.

jeherve commented:
Looking good, but you'll need to change the `@since` value from `4.0.2` to `4.1`, since `4.0.2` is already released and this filter would be implemented in `4.1`.

Thanks!

jb510 commented:
@since version bumped to 4.1.0

jb510 commented:
Note: while this does effective re-enable the publicize panel for authors without publish_posts capability, those users aren't able to actually connect to wp.com.  Looking into what's causing that and will submit another patch referencing this one.

jb510 commented:
Per the note immediately above.  We also need to filter $capability_translations ([class.JetPack.php:74](https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L74)) patch coming.

This newer (?) capability checking system isn't used by the publicize module which instead checks directly for 'publish_posts'.

Questions:
1. Instead of the filter in this PR, should the publicize module be updated to use the roles in the $capability_translations array and the filter be there?
2. If yes, should the role be Author or Contributor? it seems to me personally it should be Contributor, but historically it's been Author.

jeherve commented:
cc @dereksmart: since the Admin UI changes in 4.1, do contributors get the chance to link their account to WordPress.com now? Or is that still only available to authors?
